Sprint_2

PashaExample opened the pull request:
поправил файл и расписал проверки 
сделал фикстуры и параметризацию

PashaExample committed:
homework

AlexKitIt commented on .idea/.gitignore:
Нужно исправить: .idea это служебная папка, она должна быть в игноре. В проекте ее быть не должно

PashaExample replied:
не понял?? а где эта папка ? найти не могу в проекте?? как она туда закомитилась? где искать ее подскажи
![image](https://user-images.githubusercontent.com/129884987/232043701-8dcf321d-7eb2-489c-a126-44ec6fb8bcde.png)

PashaExample replied:
@AlexKitIt

AlexKitIt replied:
Вообще должна быть там, где ты скринишь. Надо посмотреть не отключено ли у тебя отображение. Создай .gitignore и положи туда .idea чтоб наверняка

PashaExample replied:
поправил, хотя странно , я вижу что в файле и так была эта папка

PashaExample replied:
считаю что это "МОЖНО ЛУЧШЕ" а не "НАДО ИСПРАВИТЬ"

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,138 @@
+    def books_collector(self):
+        return BooksCollector()
+
+    @pytest.fixture
+    def book1(self):
+        return "Book 1"
+
+    @pytest.fixture
+    def book2(self):
+        return "Book 2"
```
Нужно исправить: фикстура - это у нас больше про предусловие теста, а тут - просто возврат тестовых данных. Просто храни их отдельно от тестов и вызывай туда

PashaExample replied:
поправил

PashaExample replied:
убрал

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,138 @@
-    # дальше идет название метода, который тестируем add_new_book_
-    # затем, что тестируем add_two_books - добавление двух книг
-    def test_add_new_book_add_two_books(self):
+    @pytest.fixture
+    def books_collector(self):
+        return BooksCollector()
```
Нужно исправить: фикстуры лучше хранить в conftest.py

PashaExample replied:
создал файл 
перенес туда

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,138 @@
+        books_collector.add_book('Что делать, если ваш кот хочет вас убить')
+        assert len(books_collector.get_books_rating()) == 2
+
+    @pytest.mark.parametrize("rating, expected_result", [
```
Отлично: прикольная реализация, вроде и обычная параметризация, а решение с валидацией в assert неплохое

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,138 @@
+        books_collector.set_book_rating(book2, 3)
+        # Сравниваем списки слева и справа
+        assert books_collector.get_books_by_rating(rating) == expected_result
+
+
```
Можно лучше: PEP8 - нужна одна пустая строка между методами

PashaExample replied:
поправил

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,138 @@
+        (1, [])
+    ])
+    def test_get_books_by_rating(self, books_collector, book1, book2, rating, expected_result):
+        # Получение книги по рейтингу
```
Можно лучше: комменты зло. Если есть желание дать пояснение к методу - прежде всего задумайся, а правильно ли он назван, раз к нем нужно пояснение, если назван правильно а желание коммента не пропало - пиши докстринг. В твоем случае, все предельно понятно из нейминга методов, так что комменты излишни

PashaExample replied:
согласен поправил

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,138 @@
+
+        # Проверяем, что повторное добавление книги не изменяет ее рейтинг.
+        books_collector.add_book(book1)
+        assert books_collector.get_books_rating()[book1] == 1
```
Нужно исправить: для этого есть свой метод get_book_rating(book1), не стоит тут костылить и обращаться по ключам, у нас есть все готовое

PashaExample replied:
поправил

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,138 @@
+        books_collector.add_book(book1)
+        assert books_collector.get_books_rating()[book1] == 1
+
+    def test_set_book_rating(self, books_collector, book1, book2):
+        # Проверяем установку рейтинга книги.
+        books_collector.add_book(book1)
+        books_collector.set_book_rating(book1, 5)
+        assert books_collector.get_book_rating(book1) == 5
+
+        # Проверяем, что установка рейтинга не возможна для книги, которой нет в списке.
+        books_collector.set_book_rating(book2, 5)
+        assert books_collector.get_book_rating(book2) is None
+
+        # Проверяем, что рейтинг не может быть меньше 1.
+        books_collector.set_book_rating(book1, 0)
+        assert books_collector.get_book_rating(book1) == 5
+
+        # Проверяем, что рейтинг не может быть больше 10.
+        books_collector.set_book_rating(book1, 11)
+        assert books_collector.get_book_rating(book1) == 5
```
Нужно исправить: на каждый assert в ДАННОМ случае, нужен свой тест, чтобы точно знать что у нас не работает, а то завалится тест на проверке assert books_collector.get_book_rating(book1) == 5 и дальше даже не пойдет

PashaExample replied:
убрл лишние проверки) это были черновики

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,138 @@
+        books_collector.set_book_rating(book1, 11)
+        assert books_collector.get_book_rating(book1) == 5
+
+    def test_add_to_favorites(self, books_collector, book1, book2):
+        # Проверяем добавление книги в избранное.
+        books_collector.add_book(book1)
+        books_collector.add_to_favorites(book1)
+        assert book1 in books_collector.get_favorites()
+
+        # Проверяем, что книгу нельзя добавить в избранное, если её нет в словаре books_rating.
+        books_collector.add_to_favorites(book2)
+
+        # Проверяем, нет ли левого объекта (книги) в спике избранных
+        assert book2 not in books_collector.get_favorites()
```
Нужно исправить: тесты должны быть минимально-достаточны, атомарны и неделимы. Только тогда это правильные тесты, ставить несколько assert конечно можно, но в данных задачах крайне не рекомендуется. Это все-таки юниты. Один тест - одна проверка

PashaExample replied:
оставил одну проверку на один ассерт

PashaExample committed:
поправки замечаний sprint_2

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,18 @@
+    return BooksCollector()
+
+
+@pytest.fixture
+def book1():
+    return "Book 1"
+
+
+@pytest.fixture
+def book2():
+    return "Book 2"
```
Можно лучше: ты немного не понял по хранению отдельно и вызову конкретно этих методов. Это должна делать не фикстура, а обычный класс или переменная. Как фикстуру это можно использовать допустим если бы делали add_new_book(Book1) Вот это была бы фикстура, а тут просто возвращается строка Book1

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,108 @@
 from main import BooksCollector
+import pytest
+import conftest
```
Можно лучше: почитай про использование conftest и зачем этот модуль. Импортить его не надо, pytest его ищет в проекте сам

PashaExample replied:
про модуль знаю. исправлять не буду , из за того что неудобно, так более наглядно для начального уровня1

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,108 @@
+        books_collector.add_book(book_name)
+        assert book_name in books_collector.get_books_rating()
+
+    # Тут мы добавим параметризацию с двумя проверками двух книг
+    # Таким образом, при каждом вызове теста для каждой книги из book будет вызываться фикстура
+    # books_collector и создаваться объект BooksCollector()
```
Нужно исправить: ну я же писал про комменты. Не надо этого делать, это очень и очень плохая практика. Твой код должен и так понятно читаться за счет правильных неймингов, не за счет комментариев
Исправления из коммента вносить далее по коду

PashaExample replied:
поправил! считаю абсолютно глупое замечание! как новчику мне так удобнее , а ты засатвляешь убирать удобные коменты на первых этапах!!!!!

AlexKitIt replied:
Как новичку, тебе должно быть удобнее слушать, как писать свой код, чтобы потом можно было его, да и себя с ним  продать заказчику. Учиться и запоминать плохие практики - это ошибочный путь, ни понимания своей работы ни тем более ее коммерческого интереса так не получить.

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,108 @@
+    # Таким образом, при каждом вызове теста для каждой книги из book будет вызываться фикстура
+    # books_collector и создаваться объект BooksCollector()
+    @pytest.mark.parametrize("book", [
+        conftest.book1,
+        conftest.book2
```
Нужно исправить: фикстуры в параметрах теста не вызываются.  Так можно было бы сделать если бы это был простой класс и ты обращался к его атрибутам. Прямого обращения к conftest быть не должно

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,108 @@
+    def test_remove_book_that_does_not_exist(self, books_collector, book):
+        # Удаление книги из избранных, которой нет.
+        books_collector.add_book(book)
+        books_collector.remove_from_favorites(book)
```
Нужно исправить: нет такого метода, есть delete_book_from_favorites(). Код опять не запускался перед отправкой на проверку...

PashaExample replied:
поправил) тут мой косяк не запутался в ремув и делит

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,108 @@
+        # Смотрим, что рейтинг книги, а значит и сама книга остались на месте
+        assert books_collector.get_books_rating() == {book: 5}
+
+    def test_add_book(self, books_collector, book1):
+        # Проверяем добавление книги.
+        books_collector.add_book(book1)
+
+        # Проверяем, что книга есть в избранном
+        assert book1 in books_collector.get_books_rating()
```
Нужно исправить: это тест с лихвой покрыт тут test_add_new_book_add_two_books()

PashaExample replied:
поправил) ну это я доабвлял что бы , самому нагляднее было

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,108 @@
+
+    def test_set_book_rating(self, books_collector, book1, book2):
+        # Проверяем установку рейтинга книги.
+        books_collector.add_book(book1)
+        books_collector.set_book_rating(book1, 5)
+        assert books_collector.get_book_rating(book1) == 5
```
Нужно исправить: в этом тесте эти шаги лишние, поскольку set rating уже проверили здесь test_remove_book_that_does_not_exist(), и нужно проверить только присвоение рейтинга книге, которую не добавляли

PashaExample replied:
поправил ! согласен

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,108 @@
+        # Проверяем удаление книги из избранного.
+        books_collector.add_book(book1)
+        books_collector.add_to_favorites(book1)
+        books_collector.remove_from_favorites(book1)
```
Нужно исправить: уже писал, что такого метода нет

PashaExample replied:
поправил

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,108 @@
-    # напиши свои тесты ниже
-    # чтобы тесты были независимыми в каждом из них создавай отдельный экземпляр класса BooksCollector()
\ No newline at end of file
+    def test_get_books_rating(self, books_collector, book1):
+        # Получение рейтингов книг
+        books_collector.add_book(book1)
+        books_collector.set_book_rating(book1, 5)
+        assert books_collector.get_books_rating() == {book1: 5}
```
Можно лучше: методы уже проверены тут test_add_book()

PashaExample replied:
да наверное согласен!

PashaExample committed:
поправки замечаний повторные
отписался в коментариях на гит хаб по исправлениям

PashaExample committed:
поправки замечаний повторные
отписался в коментариях на гит хаб по исправлениям

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,75 @@
+        books_collector.set_book_rating(book, 5)
+        assert books_collector.get_books_rating() == {book: 5}
+
+    def test_set_book_rating(self, books_collector, book2):
+        books_collector.add_book(book2)
+        books_collector.set_book_rating(book2, 5)
+        assert books_collector.get_book_rating(book2) is None
```
Нужно исправить: логика теста неверна, при добавлении book2 метод get_book_rating() уже вернет ее с индексом 1. None там не будет

PashaExample replied:
поправил)) но в это реализации не уверен

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,75 @@
+        books_collector.add_to_favorites(book1)
+        assert len(books_collector.get_favorites()) == 1
+
+    def test_add_book_with_empty_name(self, books_collector):
+        books_collector.add_book("")
+        assert books_collector.get_books_rating() == {"": 1}
+
+    def test_add_book_with_special_characters(self, books_collector):
+        book_name = "C++ Programming for Beginners [Special Edition]\n"
+        books_collector.add_book(book_name)
+        assert book_name in books_collector.get_books_rating()
```
Можно лучше: в документации по методу нет указания на ограничения по символам, зачем два одинаковых теста?

PashaExample replied:
потому что

PashaExample replied:
мы тут учимся и я рассмотрел кейс где ограничения будут

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,75 @@
-        # проверяем, что добавилось именно две
-        # словарь books_rating, который нам возвращает метод get_books_rating, имеет длину 2
-        assert len(collector.get_books_rating()) == 2
+    @pytest.mark.parametrize("book_name, expected_result", [
+        ("", {"": 1}),
+        ("C++ Programming for Beginners [Special Edition]\n", {"C++ Programming for Beginners [Special Edition]\n": 1})
+    ])
+    def test_add_book_with_special_cases(self, books_collector, book_name, expected_result):
+        books_collector.add_book(book_name)
+        assert books_collector.get_books_rating() == expected_result
```
Нужно исправить: методы уже реализованы в test_add_book_with_empty_name() и test_add_book_with_special_characters(). Писать кучу тестов на одну и ту же функциональность -  бессмысленная трата времени разработчика, времени прогона юнитов при билде, денег заказчика на проекте

PashaExample replied:
заменил на новый

PashaExample replied:
очень жалко что все мои старания ты сливаешь в трубу

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,75 @@
+        books_collector.add_book(book2)
+        books_collector.set_book_rating(book1, 5)
+        books_collector.set_book_rating(book2, 3)
+        assert books_collector.get_books_by_rating(rating) == expected_result
```
Нужно исправить: нет такого метода, есть get_books_with_specific_rating()

PashaExample replied:
тут извиняюсь! правда недоглядел 
неопытсность

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,75 @@
-    # напиши свои тесты ниже
-    # чтобы тесты были независимыми в каждом из них создавай отдельный экземпляр класса BooksCollector()
\ No newline at end of file
+    def test_get_books_rating(self, books_collector, book1):
```
Нужно исправить: методы уже покрыты тестом test_remove_book_that_does_not_exist()

PashaExample committed:
Абсолютно с нуля переписал тесты, так как запутался в правках!
Взял список проверок из теории Я.Практикума
Обсудил нюансы тестов с куратором
сделал более простые проверки
запустил и проверил что тесты проходят
убрал коменты

p.s четсно тема далась очень сложно, но буду продолжать учиться и добивать такие тесты

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,54 @@
+        for book, book_rating in books_with_ratings.items():
+            books_collector.add_new_book(book)
+            books_collector.set_book_rating(book, book_rating)
+        assert len(books_collector.get_books_with_specific_rating(10)) == 1, 'Ни одной книги с рейтингом более десяти'
```
Нужно исправить: отсутствует проверка метода delete_book_from_favorites()

PashaExample committed:
добавил проверка метода delete_book_from_favorites()

PashaExample committed:
добавил проверка метода delete_book_from_favorites()
